typescript/sdk/src/hook/read.ts

```diff
@@ -0,0 +1,284 @@
+  protected readonly provider: providers.Provider;
+
+  constructor(protected readonly multiProvider: MultiProvider, chain: Chains) {
+    this.provider = this.multiProvider.getProvider(chain);
+  }
+
+  public static stringifyConfig(config: HookConfig, space?: number): string {
+    return JSON.stringify(
+      config,
+      (_, value) => {
+        // Check if the value looks like a serialized BigNumber
+        if (
+          typeof value === 'object' &&
+          value !== null &&
+          value.type === 'BigNumber' &&
+          value.hex
+        ) {
+          return BigNumber.from(value.hex).toString();
+        }
+        // Handle bigint values
+        return typeof value === 'bigint' ? value.toString() : value;
+      },
+      space,
+    );
+  }
+
+  async deriveHookConfig(address: Address): Promise<HookConfig> {
+    const hook = IPostDispatchHook__factory.connect(address, this.provider);
+    const onchainHookType: OnchainHookType = await hook.hookType();
+    const hookType = mapOnchainHookToHookType(onchainHookType);
+
+    switch (hookType) {
+      case HookType.MERKLE_TREE:
+        return this.deriveMerkleTreeConfig(address);
+      case HookType.AGGREGATION:
+        return this.deriveAggregationConfig(address);
+      case HookType.INTERCHAIN_GAS_PAYMASTER:
+        return this.deriveIgpConfig(address);
+      case HookType.PROTOCOL_FEE:
+        return this.deriveProtocolFeeConfig(address);
+      case HookType.OP_STACK:
+        return this.deriveOpStackConfig(address);
+      case HookType.ROUTING:
+        return this.deriveDomainRoutingConfig(address);
+      case HookType.FALLBACK_ROUTING:
+        return this.deriveFallbackRoutingConfig(address);
+      case HookType.PAUSABLE:
+        return this.derivePausableConfig(address);
+      default:
+        throw new Error(`Unsupported HookType: ${hookType}`);
+    }
+  }
+
+  async deriveMerkleTreeConfig(_: Address): Promise<MerkleTreeHookConfig> {
+    return {
+      type: HookType.MERKLE_TREE,
+    };
+  }
+
+  async deriveAggregationConfig(
+    address: Address,
+  ): Promise<AggregationHookConfig> {
+    const hook = StaticAggregationHook__factory.connect(address, this.provider);
+    const hooks = await hook.hooks(ethers.constants.AddressZero);
+    const hookConfigs = await Promise.all(
+      hooks.map(this.deriveHookConfig.bind(this)),
+    );
+
+    return {
+      type: HookType.AGGREGATION,
+      hooks: hookConfigs,
+    };
+  }
+
+  async deriveIgpConfig(address: Address): Promise<IgpHookConfig> {
+    const hook = InterchainGasPaymaster__factory.connect(
+      address,
+      this.provider,
+    );
+
+    const owner = await hook.owner();
+    const beneficiary = await hook.beneficiary();
+
+    const overhead: IgpHookConfig['overhead'] = {};
+    const oracleConfig: IgpHookConfig['oracleConfig'] = {};
+
+    let oracleKey: string | undefined;
+
+    for (const domainId of this.multiProvider.getKnownDomainIds()) {
```

Comment:
can we capture a ticket for migrating this to enumerables?

Reply:
created https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/3560, described it best as I understand but feel free to add more details